### PR TITLE
Add STDC segmentation model

### PR DESCRIPTION
**Pull request by lkdci**

### CODE REVIEW

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    transforms:
+      - _target_: super_gradients.training.utils.segmentation_utils.ColorJitterSeg
+        brightness: ${dataset_params.color_jitter}
+        contrast: ${dataset_params.color_jitter}
+        saturation: ${dataset_params.color_jitter}
+
+      - _target_: super_gradients.training.utils.segmentation_utils.RandomFlip
+
+      - _target_: super_gradients.training.utils.segmentation_utils.RandomRescale
+        scales: ${dataset_params.random_scales}
```
Comment (lkdci) [inline src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:37]: should be the random_scales above: `[0.125, 1.5]`

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  - 0.125
+  - 1.5
+  crop_size:
+    - 1024
+    - 512
+  eval_scale: 0.5
+  image_mask_transforms_aug:
+    _target_: torchvision.transforms.Compose
+    transforms:
+      - _target_: super_gradients.training.utils.segmentation_utils.ColorJitterSeg
```
Comment (lkdci) [inline src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:64]: implement __repr__ for our segmentation transforms

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  load_weights_only: True
+  strict_load: no_key_matching
+
+dataset_params:
+  _convert_: all
+  batch_size: 8
+  val_batch_size: 8
+  color_jitter: 0.5
+  random_scales:
+  - 0.125
```
Comment (lkdci) [inline src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:55]: add indent

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        saturation: ${dataset_params.color_jitter}
+
+      - _target_: super_gradients.training.utils.segmentation_utils.RandomFlip
+
+      - _target_: super_gradients.training.utils.segmentation_utils.RandomRescale
+        scales: ${dataset_params.random_scales}
+
+      - _target_: super_gradients.training.utils.segmentation_utils.PadShortToCropSize
+        crop_size: ${dataset_params.crop_size}
+        fill_mask: 19
```
Comment (lkdci) [inline src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:76]: use the CITYSCAPES_IGNORE_LABEL

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml [hunk 5]
```diff
@@ -0,0 +1,107 @@
+#  STDC segmentation training example with Cityscapes dataset.
+#  Reproduction and refinement of paper: Rethinking BiSeNet For Real-time Semantic Segmentation.
+#
+#  Usage STDC1-Seg50:
+#      python -m torch.distributed.launch --nproc_per_node=4 train_from_recipe.py --config-name=cityscapes_stdc_seg50 external_checkpoint_path=<stdc1-backbone-pretrained-path>
+#  Usage STDC2-Seg50:
+#      python -m torch.distributed.launch --nproc_per_node=4 train_from_recipe.py --config-name=cityscapes_stdc_seg50 external_checkpoint_path=<stdc2-backbone-pretrained-path> architecture=stdc2_seg
+#
```
Comment (lkdci) [inline src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:8]: Add comment about the resultion in inference

File: src/super_gradients/recipes/cityscapes_stdc_seg75.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#      STDC1-Seg75:    input-size: [768, 1536]     mIoU: 76.87     4 X RTX A5000, 29 H, early stopped after 711 epochs
+#      STDC2-Seg75:    input-size: [768, 1536]     mIoU: 78.93     2 X RTX A5000, 29 H, early stopped after 530 epochs
+#
+#  Official git repo:
+#      https://github.com/MichaelFan01/STDC-Seg
+#  Paper:
+#      https://arxiv.org/abs/2104.13188
+#
+#  Pretrained checkpoints:
+#      Backbones- downloaded from the author's official repo.
```
Comment (lkdci) [inline src/super_gradients/recipes/cityscapes_stdc_seg75.yaml:20]: Add backbone checkpoints if allowed. consider training on ImageNet.

File: src/super_gradients/recipes/training_hyperparams/cityscapes_stdc_train_params.yaml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+ema: True
+ema_params:
+  decay: 0.9999
+  beta: 15
+  exp_activation: True
+
+loss:
+  _target_: super_gradients.training.losses.stdc_loss.STDCLoss
+  num_classes: 20
```
Comment (lkdci) [inline src/super_gradients/recipes/training_hyperparams/cityscapes_stdc_train_params.yaml:23]: should be 19

File: src/super_gradients/training/models/segmentation_models/stdc.py [hunk 8]
```diff
@@ -0,0 +1,646 @@
...
+        aux_out_s32 = self.aux_head_s32(feat32)
+
+        return feat_out, aux_out_s32, aux_out_s16, detail_out8
+
+    def replace_head(self, new_num_classes: int, **kwargs):
+        ffm_channels = self.ffm.attention_block[-2].out_channels
+        dropout = self.segmentation_head[0].seg_head[1].p
+
+        self.segmentation_head = nn.Sequential(
+            SegmentationHead(ffm_channels, ffm_channels, new_num_classes, dropout=dropout),
```
Comment (lkdci) [inline src/super_gradients/training/models/segmentation_models/stdc.py:519]: you can replace only the first element in the head sequentials

File: src/super_gradients/training/losses/stdc_loss.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        losses.append(total_loss)
+
+        return total_loss, torch.stack(losses, dim=0).detach()
+
+    def get_train_named_params(self):
+        """
+        Expose DetailAggregateModule learnable parameters to be passed to the optimizer.
+        """
+        if self.use_detail:
+            return self.detail_module.named_parameters()
```
Comment (lkdci) [inline src/super_gradients/training/losses/stdc_loss.py:189]: cast to list, to be json serializable

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml (hunk 3 above)
Comment (shaydeci) [reply src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:55]: added

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml (hunk 2 above)
Comment (shaydeci) [reply src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:64]: implemented

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml (hunk 4 above)
Comment (shaydeci) [reply src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:76]: added

File: src/super_gradients/recipes/training_hyperparams/cityscapes_stdc_train_params.yaml (hunk 7 above)
Comment (shaydeci) [reply src/super_gradients/recipes/training_hyperparams/cityscapes_stdc_train_params.yaml:23]: fixed

File: src/super_gradients/training/losses/stdc_loss.py (hunk 9 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/stdc_loss.py:189]: done

File: src/super_gradients/recipes/cityscapes_stdc_seg75.yaml (hunk 6 above)
Comment (shaydeci) [reply src/super_gradients/recipes/cityscapes_stdc_seg75.yaml:20]: added

File: src/super_gradients/training/models/segmentation_models/stdc.py (hunk 8 above)
Comment (shaydeci) [reply src/super_gradients/training/models/segmentation_models/stdc.py:519]: done

File: src/super_gradients/recipes/cityscapes_stdc_seg50.yaml (hunk 5 above)
Comment (shaydeci) [reply src/super_gradients/recipes/cityscapes_stdc_seg50.yaml:8]: Resolving this comment as we said this information should be in Readme (will add it with the other benchmark info)

**Review by ofrimasad — APPROVED** [review]
LGTM

**Pull request merged**